breadbox/breadbox/api/dimension_types.py

```diff
@@ -622,24 +624,56 @@ def list_dimension_types_endpoint(db: SessionWithUser = Depends(get_db_with_user
     response_model=List[DimensionIdentifiers],
 )
 def get_dimension_type_identifiers(
+    request: Request,
     name: str,
     data_type: Annotated[Union[str, None], Query()] = None,
     show_only_dimensions_in_datasets: Annotated[bool, Query()] = False,
     limit: Annotated[Union[int, None], Query()] = None,
     db: SessionWithUser = Depends(get_db_with_user),
+    if_none_match: Optional[List[str]] = Header(None), # etag from the client's cache
 ):
+    """
+    For the given dimension type and filters, get all given IDs and labels.
+    If the `data_type` is given and/or `show_only_dimensions_in_datasets` is True, the dimension identifiers that are returned will only be those that are used within a dataset.
+    If `show_only_dimensions_in_datasets` is True, dimension identifiers within datasets, excluding the dimension type metadata, are returned
+    If neither `data_type` is given nor `show_only_dimensions_in_datasets` is True, all dimension identifiers from the given dimension type are returned.
+    """
     dim_type = type_crud.get_dimension_type(db, name)
     if dim_type is None:
         raise HTTPException(404, f"Dimension type {name} not found")
+    
+    filtered_dataset_ids: Optional[list[str]] = None
+    if data_type is not None or show_only_dimensions_in_datasets:
+        # Get subset of datasets matching the filters provided that the user has access to
+        filtered_datasets = get_datasets(
+            db,
+            db.user,
+            feature_type=dim_type.name if dim_type.axis == "feature" else None,
+            sample_type=dim_type.name if dim_type.axis == "sample" else None,
+            data_type=data_type,
+        )
+        filtered_dataset_ids = [dataset.id for dataset in filtered_datasets]
+    etag = hash_id_list([name] + (sorted(filtered_dataset_ids) if filtered_dataset_ids else []))
 
-    dimension_ids_and_labels = metadata_service.get_dimension_type_identifiers(
-        db, dim_type, data_type, show_only_dimensions_in_datasets, limit=limit,
-    )
-
-    return [
-        DimensionIdentifiers(id=id, label=label)
-        for id, label in dimension_ids_and_labels.items()
-    ]
+    # If the client already has a cached version of the data, exit early
+    if if_none_match and if_none_match[0] == etag:
+        return get_not_modified_response(etag)
+    else:
+        if filtered_dataset_ids is None:
+            dataset_ids_without_metadata = None
+        else:
+            # Remove the metadata dataset from our list of datasets 
+            dataset_ids_without_metadata = [dataset_id for dataset_id in filtered_dataset_ids if dataset_id != dim_type.dataset_id]
```

Comment:
This could be included as part of the list comprehension above correct? (I'm not sure I fully understand what this is for)